src/meta_data.rs

```diff
@@ -186,6 +173,74 @@ impl MetaData {
         }
     }
 
+    /// Get node id from request
+    fn get_node_id_from_request(&self, req: &CreateVolumeRequest) -> Option<String> {
+        if req.has_volume_content_source() {
+            let volume_src = req.get_volume_content_source();
+            if volume_src.has_snapshot() {
+                let snapshot_id = &volume_src.get_snapshot().snapshot_id;
+                let snapshot = self
+                    .get_snapshot_by_id(snapshot_id)
+                    .unwrap_or_else(|| panic!("failed to get snapshot ID={}", snapshot_id));
+                Some(snapshot.node_id)
+            } else if volume_src.has_volume() {
+                let volume_id = &volume_src.get_volume().volume_id;
+                let volume = self
+                    .get_volume_by_id(volume_id)
+                    .unwrap_or_else(|| panic!("failed to get volume ID={}", volume_id));
+                Some(volume.node_id)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Check node ID in request topology
+    fn check_node_id_in_request_topology(node_id: &str, req: &CreateVolumeRequest) -> bool {
+        if req.has_accessibility_requirements() {
+            let required_topology = req.get_accessibility_requirements().get_requisite();
+            let preferred_topology = req.get_accessibility_requirements().get_preferred();
+
+            if !required_topology
+                .iter()
+                .any(|t| t.get_segments().iter().any(|(_k, v)| v == node_id))
+                && !preferred_topology
+                    .iter()
+                    .any(|t| t.get_segments().iter().any(|(_k, v)| v == node_id))
+            {
+                return false;
+            }
+        }
+        true
+    }
+
+    /// Select a node to create volume or snapshot
+    pub fn select_node(&self, req: Option<&CreateVolumeRequest>) -> anyhow::Result<DatenLordNode> {
+        if let Some(request) = req {
```

Comment:
`CreateVolumeRequest` should always exist, no need to use `Option` here.

Reply:
Done